source/common/router/config_impl.cc

```diff
@@ -203,9 +282,13 @@ void PrefixRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) cons
   finalizePathHeader(headers, prefix_);
 }
 
-bool PrefixRouteEntryImpl::matches(const Http::HeaderMap& headers, uint64_t random_value) const {
-  return RouteEntryImplBase::matches(headers, random_value) &&
-         StringUtil::startsWith(headers.Path()->value().c_str(), prefix_, case_sensitive_);
+const Route* PrefixRouteEntryImpl::matches(const Http::HeaderMap& headers,
+                                           uint64_t random_value) const {
+  if ((RouteEntryImplBase::matches(headers, random_value) != nullptr) &&
+      StringUtil::startsWith(headers.Path()->value().c_str(), prefix_, case_sensitive_)) {
+    return this;
```

Comment:
this is where you should be returning either this, or a weighted cluster. Have a helper function that determines which one to return. Do this also in PathRouteEntryImpl::matches